proposals/0000-replacing-cabal-custom-build.md

```diff
@@ -0,0 +1,1378 @@
+  , localBuildInfo :: LocalBuildInfo
+  , benchmarkFlags :: BenchmarkFlags
+  , benchmark :: Benchmark
+  , componentLocalBuildInfo :: ComponentLocalBuildInfo
+  }
+
+type TestComponentHook      = TestComponentInputs      -> IO ()
+type BenchmarkComponentHook = BenchmarkComponentInputs -> IO ()
+
+data TestHooks
+  = TestHooks
+  { preTestComponentHook  :: Maybe TestComponentHook
+  , postTestComponentHook :: Maybe TestComponentHook
+  }
+
+data BenchmarkHooks
+  = BenchmarkHooks
+  { preBenchComponentHook  :: Maybe BenchmarkComponentHook
+  , postBenchComponentHook :: Maybe BenchmarkComponentHook
+  }
+```
+
+
+### Clean hooks
+
+Clean hooks are useful if an earlier hook has generated some additional files
+that need to be cleaned up when the user runs the build system's clean action.
+For example, the
+`Setup.hs` files [`lhs2tex-1.25`](https://hackage.haskell.org/package/lhs2tex-1.25/src/Setup.hs)
+and [`binembed-0.1.0.2`](https://hackage.haskell.org/package/binembed-0.1.0.2/docs/src/Distribution-Simple-BinEmbed.html#binembedClean)
+implement versions of this.
+
+`cabal clean` does not currently execute `./Setup.hs clean`, which appears to be a bug
+(see [#6112](https://github.com/haskell/cabal/issues/6112), [#7877](https://github.com/haskell/cabal/issues/7877)).
+In the new design we can ensure that each hook is executed before and after the
+corresponding build phase, and hence avoid such bugs.
+
+```haskell
+data CleanComponentInputs
+  = CleanComponentInputs
+  { packageDescription :: PackageDescription
+  , component :: Component
+  , cleanFlags :: CleanFlags
+  } deriving (Generic, Show)
+
+type CleanComponentHook = CleanComponentInputs -> IO ()
+
+data CleanHooks
+  = CleanHooks
+  { cleanComponentHook :: Maybe CleanComponentHook
+  }
+```
+
+
+### Pre-processors
+
+`UserHooks` includes `hookedPreProcessors`, which makes it possible to associate
+a file extension with a preprocessing operation that turns files with that
+extension into Haskell source modules.  Cabal will then execute such pre-processors
+on all matching files before building, and re-run them as needed when the input files change.
+This is similar to Cabal's built-in support for `alex`, `happy`, etc..  For example,
+[`binembed`](https://hackage.haskell.org/package/binembed-0.1.0.3/docs/src/Distribution-Simple-BinEmbed.html#withBinEmbed)
+defines a preprocessor that turns `M.binembed` into `M.hs` by running the
+`binembed` executable.
+
+A key restriction of `hookedPreProcessors` is that it assumes one input file
+will turn into one Haskell module (see [#6232 Extended preprocessors
+support](https://github.com/haskell/cabal/issues/6232)).
+
+Packages requiring pre-processors could define a pre-build hook that searches
+for files with their custom extension and produces the corresponding Haskell
+file.  However this would require duplicating logic present in `Cabal`, and lead
+to worse recompilation behaviour.  Thus we propose to retain
```

Comment:
Yes adding lots of extraneous domain-specific concepts because we do not understand what hooks are supposed to be doing is bad imperative thought, and will garbage up the implementation.

This is why I am advocating fine-grained hooks from the get-go. It is a much simpler conceptual model that completely avoids these conceptual difficulties.